pypy failures

Going through the logs from @mattip 's logs in #19154 I think the failures we are seeing can be classified as below.


## finalizer / weakrefs different in pypy

We probably should just detect and skip pypy on these.

- [ ] FAILED lib/matplotlib/tests/test_animation.py::test_animation_delete[anim0]
- [ ] FAILED lib/matplotlib/tests/test_animation.py::test_animation_repr_html[anim0-none-None-imagemagick]
- [ ] FAILED lib/matplotlib/tests/test_animation.py::test_animation_repr_html[anim0-none-None-ffmpeg]
- [ ] FAILED lib/matplotlib/tests/test_animation.py::test_funcanimation_cache_frame_data[False]
- [ ] FAILED lib/matplotlib/tests/test_cbook.py::Test_callback_registry::test_callback_complete

## pypy does not implement a working `sys.getsizeof`

Skip this test on pypy, we are only using it to make sure that the output of rendering makes sense.

- [x] FAILED lib/matplotlib/tests/test_backend_pdf.py::test_multipage_properfinalize


## stack walking different?

The failures due to raising an unexpected warning are a bit more concerning.  We are doing some frame walking / inspection to sort out if we should warn or not in https://github.com/matplotlib/matplotlib/blob/09ab2c280bd7abb07b3e35a39134a9f791781fc9/lib/matplotlib/backend_bases.py#L1624-L1665  I assume this is not working on pypy because we are relying on some cpython implementation detail?  

If someone who understands pypy can fix that to work on both that would be best, but if not we should detect if we are on pypy and fall back to never warning.  It is operating in a slightly degraded mode (users may not get warnings they should see), but the status quo is going to warn for things the user can not do anything about (I think every time we use tightlayout of `bbox_inches='tight'`).

This is probably the only failures I would hold 3.4 over.

- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight[png]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight[svg]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_suptile_legend[pdf]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_clipping[png]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_clipping[svg]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_raster[pdf]
- [ ] FAILED lib/matplotlib/tests/test_backend_pgf.py::test_bbox_inches_tight - mat...
- [ ] FAILED lib/matplotlib/tests/test_backend_pgf.py::test_mixedmode[pdf] - matplo...
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight[pdf]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_suptile_legend[png]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_suptile_legend[svg]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_clipping[pdf]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_raster[png]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_bbox_inches_tight_raster[svg]
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_tight_pcolorfast - matpl...
- [ ] FAILED lib/matplotlib/tests/test_figure.py::test_tightbbox_box_aspect[svg] - ...
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_only_on_non_finite_bbox
- [ ] FAILED lib/matplotlib/tests/test_bbox_tight.py::test_noop_tight_bbox - matplo...
- [ ] FAILED lib/matplotlib/tests/test_offsetbox.py::test_annotationbbox_extents - ...
- [ ] FAILED lib/mpl_toolkits/tests/test_axes_grid1.py::test_image_grid[png] - matp...

## segfaults in tk

We should either track this down or just skip the tests.

- [ ] FAILED lib/matplotlib/tests/test_backends_interactive.py::test_interactive_backend[toolbar2-tkagg]
- [ ] FAILED lib/matplotlib/tests/test_backends_interactive.py::test_interactive_backend[toolmanager-gtk3agg]
- [ ] FAILED lib/matplotlib/tests/test_backends_interactive.py::test_interactive_backend[toolmanager-tkagg]


## subprocess hung

We should either track this down or just skip the tests.


- [ ] FAILED lib/matplotlib/tests/test_backend_tk.py::test_never_update - Failed: S...
- [ ] FAILED lib/matplotlib/tests/test_backend_tk.py::test_figuremanager_cleans_own_mainloop
- [ ] FAILED lib/matplotlib/tests/test_backends_interactive.py::test_interactive_backend[toolbar2-gtk3agg]
- [ ] FAILED lib/matplotlib/tests/test_backends_interactive.py::test_interactive_thread_safety[gtk3agg]
- [ ] FAILED lib/matplotlib/tests/test_backends_interactive.py::test_interactive_thread_safety[tkagg]

## not matching numpy?

These both look like they are upstream bugs in pypy?  We should probably just skip these. 

- [ ] FAILED lib/matplotlib/tests/test_cbook.py::test_step_fails[args1] - SystemErr...
- [ ] FAILED lib/matplotlib/tests/test_matplotlib.py::test_importable_with__OO - As...

## checking refcounts

These tests do not make sense for pypy, we should skip them.

- [ ] FAILED lib/matplotlib/tests/test_quiver.py::test_quiver_memory_leak - Attribu...
- [ ] FAILED lib/matplotlib/tests/test_quiver.py::test_quiver_key_memory_leak - Att...

## sockets not working

The comments around this suggest we never got this working on azure (which I assume is what is hosting GH actions) so we should sort out why the code to not run these tests in that case is not being triggered here.

- [ ] FAILED lib/matplotlib/tests/test_backends_interactive.py::test_webagg - Faile...

I don't know how to fix these animation warning tests, because adding `gc.collect` does not help:
```
FAILED lib/matplotlib/tests/test_animation.py::test_animation_delete[anim0] - Failed: DID NOT WARN. No warnings of type (<class 'Warning'>,) was emitted. The list of emitted warnings is: [].
FAILED lib/matplotlib/tests/test_animation.py::test_animation_repr_html[anim0-none-None-ffmpeg] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) was emitted. The list of emitted warnings is: [].
FAILED lib/matplotlib/tests/test_animation.py::test_animation_repr_html[anim0-none-None-imagemagick] - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) was emitted. The list of emitted warnings is: [].
```
Clearly, the warning is happening, because it is printed out to stderr:
```
Exception ignored in: <bound method Animation.__del__ of <matplotlib.animation.FuncAnimation object at 0x0000000004839210>>
Traceback (most recent call last):
  File ".../lib/matplotlib/animation.py", line 890, in __del__
    'Animation was deleted without rendering anything. This is '
UserWarning: Animation was deleted without rendering anything. This is most likely unintended. To prevent deletion, assign the Animation to a variable that exists for as long as you need the Animation.
```
but it does not seem to get captures by `pytest.warns`.

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -47,7 +47,7 @@ jobs:
 
       - name: Install cibuildwheel
         run: |
-          python -m pip install cibuildwheel==1.9.0
+          python -m pip install cibuildwheel==2.1.1
 
       - name: Build minimum NumPy for aarch64
         if: matrix.cibw_archs == 'aarch64' && steps.numpy-cache.outputs.cache-hit != 'true'

--- a/.github/workflows/cibuildwheel.yml
+++ b/.github/workflows/cibuildwheel.yml
@@ -93,7 +93,7 @@ jobs:
           CIBW_BUILD: "pp37-*"
           CIBW_BEFORE_BUILD: pip install certifi numpy==${{ env.min-numpy-version }}
           CIBW_ARCHS: ${{ matrix.cibw_archs }}
-        if: runner.os != 'Windows' && matrix.cibw_archs != 'aarch64'
+        if: matrix.cibw_archs != 'aarch64'
 
       - name: Validate that LICENSE files are included in wheels
         run: |

--- a/lib/matplotlib/backend_bases.py
+++ b/lib/matplotlib/backend_bases.py
@@ -1618,14 +1618,20 @@ def _check_savefig_extra_args(func=None, extra_kwargs=()):
             if frame is None:
                 # when called in embedded context may hit frame is None.
                 break
+            # Work around sphinx-gallery not setting __name__.
+            frame_name = frame.f_globals.get('__name__', '')
             if re.match(r'\A(matplotlib|mpl_toolkits)(\Z|\.(?!tests\.))',
-                        # Work around sphinx-gallery not setting __name__.
-                        frame.f_globals.get('__name__', '')):
-                if public_api.match(frame.f_code.co_name):
-                    name = frame.f_code.co_name
+                        frame_name):
+                name = frame.f_code.co_name
+                if public_api.match(name):
                     if name in ('print_figure', '_no_output_draw'):
                         seen_print_figure = True
 
+            elif frame_name == '_functools':
+                # PyPy adds an extra frame without module prefix for this
+                # functools wrapper, which we ignore to assume we're still in
+                # Matplotlib code.
+                continue
             else:
                 break
 

--- a/lib/matplotlib/patches.py
+++ b/lib/matplotlib/patches.py
@@ -1451,7 +1451,8 @@ class FancyArrow(Polygon):
 
 
 docstring.interpd.update(
-    FancyArrow="\n".join(inspect.getdoc(FancyArrow.__init__).splitlines()[2:]))
+    FancyArrow="\n".join(
+        (inspect.getdoc(FancyArrow.__init__) or "").splitlines()[2:]))
 
 
 class CirclePolygon(RegularPolygon):

--- a/lib/matplotlib/scale.py
+++ b/lib/matplotlib/scale.py
@@ -620,10 +620,11 @@ def _get_scale_docs():
     """
     docs = []
     for name, scale_class in _scale_mapping.items():
+        docstring = inspect.getdoc(scale_class.__init__) or ""
         docs.extend([
             f"    {name!r}",
             "",
-            textwrap.indent(inspect.getdoc(scale_class.__init__), " " * 8),
+            textwrap.indent(docstring, " " * 8),
             ""
         ])
     return "\n".join(docs)

--- a/lib/matplotlib/tests/test_animation.py
+++ b/lib/matplotlib/tests/test_animation.py
@@ -1,5 +1,6 @@
 import os
 from pathlib import Path
+import platform
 import subprocess
 import sys
 import weakref
@@ -86,6 +87,11 @@ def test_null_movie_writer(anim):
 
 @pytest.mark.parametrize('anim', [dict(klass=dict)], indirect=['anim'])
 def test_animation_delete(anim):
+    if platform.python_implementation() == 'PyPy':
+        # Something in the test setup fixture lingers around into the test and
+        # breaks pytest.warns on PyPy. This garbage collection fixes it.
+        # https://foss.heptapod.net/pypy/pypy/-/issues/3536
+        np.testing.break_cycles()
     anim = animation.FuncAnimation(**anim)
     with pytest.warns(Warning, match='Animation was deleted'):
         del anim
@@ -200,6 +206,11 @@ def test_save_animation_smoketest(tmpdir, writer, frame_format, output, anim):
 ])
 @pytest.mark.parametrize('anim', [dict(klass=dict)], indirect=['anim'])
 def test_animation_repr_html(writer, html, want, anim):
+    if platform.python_implementation() == 'PyPy':
+        # Something in the test setup fixture lingers around into the test and
+        # breaks pytest.warns on PyPy. This garbage collection fixes it.
+        # https://foss.heptapod.net/pypy/pypy/-/issues/3536
+        np.testing.break_cycles()
     if (writer == 'imagemagick' and html == 'html5'
             # ImageMagick delegates to ffmpeg for this format.
             and not animation.FFMpegWriter.isAvailable()):

--- a/lib/matplotlib/tests/test_animation.py
+++ b/lib/matplotlib/tests/test_animation.py
@@ -1,4 +1,3 @@
-import gc
 import os
 from pathlib import Path
 import subprocess
@@ -90,7 +89,7 @@ def test_animation_delete(anim):
     anim = animation.FuncAnimation(**anim)
     with pytest.warns(Warning, match='Animation was deleted'):
         del anim
-        gc.collect()
+        np.testing.break_cycles()
 
 
 def test_movie_writer_dpi_default():
@@ -214,7 +213,8 @@ def test_animation_repr_html(writer, html, want, anim):
     if want is None:
         assert html is None
         with pytest.warns(UserWarning):
-            del anim  # Animtion was never run, so will warn on cleanup.
+            del anim  # Animation was never run, so will warn on cleanup.
+            np.testing.break_cycles()
     else:
         assert want in html
 
@@ -324,6 +324,7 @@ def test_funcanimation_cache_frame_data(cache_frame_data):
     writer = NullMovieWriter()
     anim.save('unused.null', writer=writer)
     assert len(frames_generated) == 5
+    np.testing.break_cycles()
     for f in frames_generated:
         # If cache_frame_data is True, then the weakref should be alive;
         # if cache_frame_data is False, then the weakref should be dead (None).

--- a/lib/matplotlib/tests/test_backend_tk.py
+++ b/lib/matplotlib/tests/test_backend_tk.py
@@ -1,6 +1,7 @@
 import functools
 import inspect
 import os
+import platform
 import re
 import subprocess
 import sys
@@ -111,6 +112,9 @@ def test_figuremanager_preserves_host_mainloop():  # pragma: no cover
         print("success")
 
 
+@pytest.mark.skipif(platform.python_implementation() != 'CPython',
+                    reason='PyPy does not support Tkinter threading: '
+                           'https://foss.heptapod.net/pypy/pypy/-/issues/1929')
 @pytest.mark.backend('TkAgg', skip_on_importerror=True)
 @pytest.mark.flaky(reruns=3)
 @_isolated_tk_test(success_count=1)

--- a/lib/matplotlib/tests/test_backends_interactive.py
+++ b/lib/matplotlib/tests/test_backends_interactive.py
@@ -3,6 +3,7 @@ import importlib.util
 import inspect
 import json
 import os
+import platform
 import signal
 import subprocess
 import sys
@@ -220,6 +221,12 @@ for param in _thread_safe_backends:
     elif param.values[0].get("QT_API") == "PySide2":
         param.marks.append(
             pytest.mark.xfail(raises=subprocess.CalledProcessError))
+    elif backend == "tkagg" and platform.python_implementation() != 'CPython':
+        param.marks.append(
+            pytest.mark.xfail(
+                reason='PyPy does not support Tkinter threading: '
+                       'https://foss.heptapod.net/pypy/pypy/-/issues/1929',
+                strict=True))
 
 
 @pytest.mark.parametrize("env", _thread_safe_backends)

--- a/lib/matplotlib/tests/test_cbook.py
+++ b/lib/matplotlib/tests/test_cbook.py
@@ -198,11 +198,13 @@ class Test_callback_registry:
         return count1
 
     def is_empty(self):
+        np.testing.break_cycles()
         assert self.callbacks._func_cid_map == {}
         assert self.callbacks.callbacks == {}
         assert self.callbacks._pickled_cids == set()
 
     def is_not_empty(self):
+        np.testing.break_cycles()
         assert self.callbacks._func_cid_map != {}
         assert self.callbacks.callbacks != {}
 

--- a/lib/matplotlib/tests/test_quiver.py
+++ b/lib/matplotlib/tests/test_quiver.py
@@ -1,6 +1,9 @@
+import platform
+import sys
+
 import numpy as np
 import pytest
-import sys
+
 from matplotlib import pyplot as plt
 from matplotlib.testing.decorators import image_comparison
 
@@ -15,6 +18,8 @@ def draw_quiver(ax, **kw):
     return Q
 
 
+@pytest.mark.skipif(platform.python_implementation() != 'CPython',
+                    reason='Requires CPython')
 def test_quiver_memory_leak():
     fig, ax = plt.subplots()
 
@@ -27,6 +32,8 @@ def test_quiver_memory_leak():
     assert sys.getrefcount(ttX) == 2
 
 
+@pytest.mark.skipif(platform.python_implementation() != 'CPython',
+                    reason='Requires CPython')
 def test_quiver_key_memory_leak():
     fig, ax = plt.subplots()
 

--- a/lib/matplotlib/tests/test_style.py
+++ b/lib/matplotlib/tests/test_style.py
@@ -1,9 +1,9 @@
 from contextlib import contextmanager
-import gc
 from pathlib import Path
 from tempfile import TemporaryDirectory
 import sys
 
+import numpy as np
 import pytest
 
 import matplotlib as mpl
@@ -165,7 +165,7 @@ def test_xkcd_no_cm():
     assert mpl.rcParams["path.sketch"] is None
     plt.xkcd()
     assert mpl.rcParams["path.sketch"] == (1, 100, 2)
-    gc.collect()
+    np.testing.break_cycles()
     assert mpl.rcParams["path.sketch"] == (1, 100, 2)
 
 

--- a/src/_c_internal_utils.c
+++ b/src/_c_internal_utils.c
@@ -68,7 +68,13 @@ mpl_GetCurrentProcessExplicitAppUserModelID(PyObject* module)
     wchar_t* appid = NULL;
     HRESULT hr = GetCurrentProcessExplicitAppUserModelID(&appid);
     if (FAILED(hr)) {
+#if defined(PYPY_VERSION_NUM) && PYPY_VERSION_NUM < 0x07030600
+        /* Remove when we require PyPy 7.3.6 */
+        PyErr_SetFromWindowsErr(hr);
+        return NULL;
+#else
         return PyErr_SetFromWindowsErr(hr);
+#endif
     }
     PyObject* py_appid = PyUnicode_FromWideChar(appid, -1);
     CoTaskMemFree(appid);
@@ -89,7 +95,13 @@ mpl_SetCurrentProcessExplicitAppUserModelID(PyObject* module, PyObject* arg)
     HRESULT hr = SetCurrentProcessExplicitAppUserModelID(appid);
     PyMem_Free(appid);
     if (FAILED(hr)) {
+#if defined(PYPY_VERSION_NUM) && PYPY_VERSION_NUM < 0x07030600
+        /* Remove when we require PyPy 7.3.6 */
+        PyErr_SetFromWindowsErr(hr);
+        return NULL;
+#else
         return PyErr_SetFromWindowsErr(hr);
+#endif
     }
     Py_RETURN_NONE;
 #else

